File: PRACTICAS/CAMBIAR_RUBY/RUBY2/Civitas/pkg/RubyApplication1-0.0.1/lib/Casilla.rb
module Civitas
  class Casilla
    attr_accessor :nombre, :tipo, :importe, :numCasillaCarcel, :titulo, :mazo
    def initialize (n, tipo, importe, numCasillaCarcel, titulo, mazo)
      init()
      @nombre = n
      @tipo = tipo
      @importe = importe
      @numCasillaCarcel = numCasillaCarcel
      @titulo = titulo
      @mazo = mazo
      @sorpresa
    end
    
    def init
      @nombre = "nombre"
      @importe = 0
      @carcel = 0
    end
    
    def informe(actual, todos)
      Diario.instance.ocurre_evento(todos[actual].nombre + " ha caido en la casilla: #{nombre}")
      puts Diario.instance.leer_evento
      puts toString
    end
    
    def toString
      return "nombre: #{nombre} - importe: #{importe} - carcel: #{numCasillaCarcel}"
    end
    
    def jugadorCorrecto(actual, todos)
      return actual < todos.lenght
    end
    
    def recibeJugador_impuesto(actual, todos)
      if (jugadorCorrecto(actual, todos))
        informe(actual, todos)
        todos[actual].paga(@importe)
      end
    end
    
    def recibeJugador_juez(actual, todos)
      if (jugadorCorrecto(actual, todos))
        informe(actual, todos)
        todos[actual].encarcelar(@numCasillaCarcel)
      end
    end
    
    def recibeJugador(actual, todos)
      case @tipo
      when TipoCasilla::CALLE
        recibeJugador_calle(actual,todos)
      when TipoCasilla::IMPUESTO
        recibeJugador_impuesto(actual,todos)
      when TipoCasilla::JUEZ
        recibeJugador_juez(actual, todos)
      when TipoCasilla::SORPRESA
        recibeJugador_sorpresa(actual,todos)
      else
        informe(actual,todos)
      end
    end
    
    def recibeJugador_calle(actual, todos)
      if (jugadorCorrecto(actual, todos))
        informe(actual, todos)
        jugador = todos[actual]
        if (!@titulo.tienePropietario)
          jugador.puedeComprarCasilla
        else
          @titulo.tramitarAlquiler(jugador)
        end
      end
    end
    
    def recibeJugador_sorpresa(actual, todos)
      if (jugadorCorrecto(actual, todos))
        sorpresa = @mazo.siguiente
        informe(actual,todos)
        sorpresa.aplicarAJugador(actual, todos)
      end
    end
    
    private :init
  end
end

File: PRACTICAS/practica3/Civitas_P1/src/civitas_p1/Casilla.java
/*
 * To change this license header, choose License Headers in Project Properties.
 * To change this template file, choose Tools | Templates
 * and open the template in the editor. eeeeeeeeeeeeeee
 */
package civitas_p1;

import java.util.ArrayList;

/**
 *
 * @author angelsolano
 */
public class Casilla {
    
    private String nombre;
    private static int carcel;
    private float importe;
    private TipoCasilla tipo;
    private TituloPropiedad tituloPropiedad;
    private MazoSorpresas mazo;
    private Sorpresa sorpresa;
    
   
    Casilla(String nombre){
        
        init();
        this.nombre = nombre;
        tipo = TipoCasilla.DESCANSO;
        
    }
    
    Casilla(TituloPropiedad titulo){
        init();
        tituloPropiedad = titulo;
        nombre = titulo.getNombre();
        tipo = TipoCasilla.CALLE;       
        importe = titulo.getPrecioCompra();
    }
    
    Casilla(float cantidad, String nombre){
        init();
        this.nombre = nombre;
        importe = cantidad;
        tipo = TipoCasilla.IMPUESTO;
    }
    
    Casilla(int numCasillaCarcel, String nombre){
        init();
        this.nombre = nombre;
        carcel = numCasillaCarcel;
        tipo = TipoCasilla.JUEZ;
    }
    
    Casilla(MazoSorpresas mazo, String nombre){
        init();
        this.nombre = nombre;
        this.mazo = mazo;
        tipo = TipoCasilla.SORPRESA;
    }
    
    private void init(){
        nombre = "UNDEFINED";
        carcel = 8;
        importe = (float) 0.0;
        tituloPropiedad = new TituloPropiedad("UNDEFINED",0,0,0,0,0);
        mazo = null;
        sorpresa = null;
        
        
    }
    
    public String getNombre(){
        return nombre;
    }
    
    private void informe(int iactual, ArrayList<Jugador> todos){
        
        Diario.instance.ocurreEvento("El jugador " + todos.get(iactual).getNombre() + " ha caido en la casilla: " + toString());
        
    }
    
    void recibeJugador(int iactual, ArrayList<Jugador> todos){

        switch (tipo) {
            case CALLE:
                recibeJugador_calle(iactual, todos);
                
                break;
            case IMPUESTO:
                recibeJugador_impuesto(iactual, todos);
                break;
            case JUEZ:
                recibeJugador_juez(iactual, todos);
                break;
            case SORPRESA:
                recibeJugador_sorpresa(iactual, todos);
                break;
            default:
                informe(iactual, todos);
                break;
        }

        
    }
    
    private void recibeJugador_calle(int iactual, ArrayList<Jugador> todos){
        
        if(jugadorCorrecto(iactual, todos)){
            
            informe(iactual, todos);
                  
            if(!tituloPropiedad.tienePropietario()){
                
                todos.get(iactual).puedeComprarCasilla();
            }else{
                
                tituloPropiedad.tramitarAlquiler(todos.get(iactual));
                      
                
            }
           
        }
       
    }
    
    
    
    private void recibeJugador_impuesto(int iactual, ArrayList<Jugador> todos){
        if (jugadorCorrecto(iactual, todos)){
            informe(iactual, todos);
            todos.get(iactual).pagaImpuesto(importe);
        }
    }
    
    private void recibeJugador_juez(int iactual, ArrayList<Jugador> todos){
        if (jugadorCorrecto(iactual, todos)){
            informe(iactual, todos);
            todos.get(iactual).encarcelar(carcel);
        }              
    }
    
    private void recibeJugador_sorpresa(int iactual, ArrayList<Jugador> todos){
    
        if(jugadorCorrecto(iactual,todos)){

            sorpresa = mazo.siguiente();
            informe(iactual,todos);
            sorpresa.aplicarAJugador(iactual, todos);

        }
        
    }
    
    public boolean jugadorCorrecto(int iactual, ArrayList<Jugador> todos){
        return iactual >= 0 && iactual < todos.size();
    }
    
    @Override
    public String toString(){   
        return nombre + ", la carcel está en la posición " + carcel + ", el importe es de " + importe + " y es de tipo " + tipo + ".";
    }
    
    public TituloPropiedad getTituloPropiedad(){
        return tituloPropiedad;
    }
    
    public TipoCasilla getTipo(){
        return tipo;
    }
    
    
    
}
